File: StackArithmetic.java
import java.util.*;

class Stack {
    private ArrayList<Double> items;
    
    public Stack() {
        items = new ArrayList<>();
    }
    
    public void push(double item) {
        items.add(item);
    }
    
    public double pop() {
        if (!isEmpty()) {
            return items.remove(items.size() - 1);
        }
        return Double.NaN;
    }
    
    public boolean isEmpty() {
        return items.isEmpty();
    }
    
    public void clear() {
        items.clear();
    }
    
    public String display() {
        if (items.isEmpty()) return "[]";
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < items.size(); i++) {
            sb.append(formatAngka(items.get(i)));
            if (i < items.size() - 1) sb.append(", ");
        }
        sb.append("]");
        return sb.toString();
    }
    
    private String formatAngka(double nilai) {
        if (nilai == (long) nilai) {
            return String.valueOf((long) nilai);
        } else {
            return String.valueOf(nilai);
        }
    }
}

public class StackArithmetic {
    
    private static int prioritas(String op) {
        switch (op) {
            case "+": case "-": return 1;
            case "*": case "/": return 2;
            case "^": return 3;
            default: return 0;
        }
    }
    
    private static boolean isOperator(String token) {
        return token.equals("+") || token.equals("-") || 
               token.equals("*") || token.equals("/") || token.equals("^");
    }
    
    private static boolean isOperand(String token) {
        try {
            Double.parseDouble(token);
            return true;
        } catch (NumberFormatException e) {
            return false;
        }
    }
    
    private static String formatAngka(double nilai) {
        if (nilai == (long) nilai) {
            return String.valueOf((long) nilai);
        } else {
            return String.valueOf(nilai);
        }
    }
    
    private static String buatGaris(String teks) {
        StringBuilder garis = new StringBuilder();
        for (int i = 0; i < teks.length(); i++) {
            garis.append("-");
        }
        return garis.toString();
    }
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        String judul = "STACK ARITHMETIC";
        int lebar = judul.length();
        String garis = buatGaris(judul);
        
        // Judul di tengah
        System.out.println(judul);
        System.out.println(garis);
        System.out.println("Masukkan ekspresi (pisahkan spasi)");
        System.out.println("Contoh: ( 3 + 5 ) * 2 - 8 / 4");
        System.out.println("Angka: 2, 2.5, 2.50, 2.05, .5");
        System.out.println(garis);
        System.out.print("Ekspresi: ");
        
        String infix = scanner.nextLine();
        String[] tokens = infix.split(" ");
        
        // Validasi dasar
        boolean valid = true;
        for (int i = 0; i < tokens.length; i++) {
            if (!isOperand(tokens[i]) && !isOperator(tokens[i]) && 
                !tokens[i].equals("(") && !tokens[i].equals(")")) {
                System.out.println("\nError: Token '" + tokens[i] + "' tidak dikenal");
                valid = false;
                break;
            }
            
            if (i > 0 && isOperator(tokens[i]) && isOperator(tokens[i-1])) {
                System.out.println("\nError: Operator berurutan '" + tokens[i-1] + " " + tokens[i] + "'");
                valid = false;
                break;
            }
            
            if (i > 0 && isOperand(tokens[i]) && isOperand(tokens[i-1])) {
                System.out.println("\nError: Angka berurutan tanpa operator");
                valid = false;
                break;
            }
        }
        
        if (!valid) {
            System.out.println(garis);
            scanner.close();
            return;
        }
        
        // KONVERSI INFIX KE POSTFIX
        String header1 = "--- KONVERSI INFIX KE POSTFIX ---";
        System.out.println("\n" + header1);
        System.out.println(buatGaris(header1));
        
        java.util.Stack<String> opStack = new java.util.Stack<>();
        ArrayList<String> output = new ArrayList<>();
        
        for (String token : tokens) {
            if (isOperand(token)) {
                output.add(token);
                System.out.println("Token: " + token + " (angka) -> output");
            }
            else if (token.equals("(")) {
                opStack.push(token);
                System.out.println("Token: ( -> push ke stack");
            }
            else if (token.equals(")")) {
                while (!opStack.isEmpty() && !opStack.peek().equals("(")) {
                    output.add(opStack.pop());
                }
                opStack.pop(); // buang '('
                System.out.println("Token: ) -> pop stack sampai ketemu (");
            }
            else if (isOperator(token)) {
                while (!opStack.isEmpty() && isOperator(opStack.peek()) && 
                       prioritas(opStack.peek()) >= prioritas(token)) {
                    output.add(opStack.pop());
                }
                opStack.push(token);
                System.out.println("Token: " + token + " -> push ke stack");
            }
            
            System.out.println("  Output: " + output + ", Stack: " + opStack);
        }
        
        while (!opStack.isEmpty()) {
            output.add(opStack.pop());
        }
        
        String postfix = String.join(" ", output);
        System.out.println("\nHasil Postfix: " + postfix);
        
        // EVALUASI POSTFIX
        String header2 = "--- EVALUASI POSTFIX ---";
        System.out.println("\n" + header2);
        System.out.println(buatGaris(header2));
        
        Stack evalStack = new Stack();
        String[] postTokens = postfix.split(" ");
        
        for (String token : postTokens) {
            if (isOperand(token)) {
                double nilai = Double.parseDouble(token);
                evalStack.push(nilai);
                System.out.println("Token: " + token + " (angka) -> push " + formatAngka(nilai));
            }
            else if (isOperator(token)) {
                double b = evalStack.pop();
                double a = evalStack.pop();
                double hasil = 0;
                String simbol = "";
                
                switch (token) {
                    case "+": hasil = a + b; simbol = "+"; break;
                    case "-": hasil = a - b; simbol = "-"; break;
                    case "*": hasil = a * b; simbol = "×"; break;
                    case "/": hasil = a / b; simbol = "÷"; break;
                    case "^": hasil = Math.pow(a, b); simbol = "^"; break;
                }
                
                evalStack.push(hasil);
                System.out.println("Token: " + token + " (operator)");
                System.out.println("  Pop " + formatAngka(a) + " dan " + formatAngka(b));
                System.out.println("  " + formatAngka(a) + " " + simbol + " " + formatAngka(b) + 
                                 " = " + formatAngka(hasil));
            }
            
            System.out.println("  Stack: " + evalStack.display());
        }
        
        double hasilAkhir = evalStack.pop();
        System.out.println("\nHasil Akhir: " + formatAngka(hasilAkhir));
        System.out.println(garis);
        
        scanner.close();
    }
}

File: StackArithmetic.py
class Stack:
    def __init__(self):
        self.items = []
    
    def push(self, item):
        self.items.append(item)
    
    def pop(self):
        if not self.is_empty():
            return self.items.pop()
        return None
    
    def is_empty(self):
        return len(self.items) == 0
    
    def clear(self):
        self.items = []
    
    def display(self):
        return [self._format_angka(x) for x in self.items]
    
    def _format_angka(self, nilai):
        if nilai.is_integer():
            return int(nilai)
        return nilai

def prioritas(op):
    if op in ('+', '-'):
        return 1
    if op in ('*', '/'):
        return 2
    if op == '^':
        return 3
    return 0

def is_operator(token):
    return token in ('+', '-', '*', '/', '^')

def is_operand(token):
    try:
        float(token)
        return True
    except ValueError:
        return False

def format_angka(nilai):
    if nilai.is_integer():
        return str(int(nilai))
    else:
        # Hilangkan trailing zeros
        s = str(nilai)
        if '.' in s:
            s = s.rstrip('0').rstrip('.')
        return s

def buat_garis(teks):
    return "-" * len(teks)

def main():
    judul = "STACK ARITHMETIC"
    garis = buat_garis(judul)
    
    # Judul di tengah
    print(judul)
    print(garis)
    print("Masukkan ekspresi (pisahkan spasi)")
    print("Contoh: ( 3 + 5 ) * 2 - 8 / 4")
    print("Angka: 2, 2.5, 2.50, 2.05, .5")
    print(garis)
    
    infix = input("Ekspresi: ")
    tokens = infix.split()
    
    # Validasi dasar
    valid = True
    for i, token in enumerate(tokens):
        if not (is_operand(token) or is_operator(token) or token in ('(', ')')):
            print(f"\nError: Token '{token}' tidak dikenal")
            valid = False
            break
        
        if i > 0 and is_operator(token) and is_operator(tokens[i-1]):
            print(f"\nError: Operator berurutan '{tokens[i-1]} {token}'")
            valid = False
            break
        
        if i > 0 and is_operand(token) and is_operand(tokens[i-1]):
            print("\nError: Angka berurutan tanpa operator")
            valid = False
            break
    
    if not valid:
        print(garis)
        return
    
    # KONVERSI INFIX KE POSTFIX
    header1 = "--- KONVERSI INFIX KE POSTFIX ---"
    print(f"\n{header1}")
    print(buat_garis(header1))
    
    op_stack = []
    output = []
    
    for token in tokens:
        if is_operand(token):
            output.append(token)
            print(f"Token: {token} (angka) -> output")
        elif token == '(':
            op_stack.append(token)
            print("Token: ( -> push ke stack")
        elif token == ')':
            while op_stack and op_stack[-1] != '(':
                output.append(op_stack.pop())
            op_stack.pop()  # buang '('
            print("Token: ) -> pop stack sampai ketemu (")
        elif is_operator(token):
            while (op_stack and is_operator(op_stack[-1]) and 
                   prioritas(op_stack[-1]) >= prioritas(token)):
                output.append(op_stack.pop())
            op_stack.append(token)
            print(f"Token: {token} -> push ke stack")
        
        print(f"  Output: {output}, Stack: {op_stack}")
    
    while op_stack:
        output.append(op_stack.pop())
    
    postfix = ' '.join(output)
    print(f"\nHasil Postfix: {postfix}")
    
    # EVALUASI POSTFIX
    header2 = "--- EVALUASI POSTFIX ---"
    print(f"\n{header2}")
    print(buat_garis(header2))
    
    eval_stack = Stack()
    post_tokens = postfix.split()
    
    for token in post_tokens:
        if is_operand(token):
            nilai = float(token)
            eval_stack.push(nilai)
            print(f"Token: {token} (angka) -> push {format_angka(nilai)}")
        elif is_operator(token):
            b = eval_stack.pop()
            a = eval_stack.pop()
            
            if token == '+':
                hasil = a + b
                simbol = '+'
            elif token == '-':
                hasil = a - b
                simbol = '-'
            elif token == '*':
                hasil = a * b
                simbol = '×'
            elif token == '/':
                hasil = a / b
                simbol = '÷'
            elif token == '^':
                hasil = a ** b
                simbol = '^'
            
            eval_stack.push(hasil)
            print(f"Token: {token} (operator)")
            print(f"  Pop {format_angka(a)} dan {format_angka(b)}")
            print(f"  {format_angka(a)} {simbol} {format_angka(b)} = {format_angka(hasil)}")
        
        print(f"  Stack: {eval_stack.display()}")
    
    hasil_akhir = eval_stack.pop()
    print(f"\nHasil Akhir: {format_angka(hasil_akhir)}")
    print(garis)

if __name__ == "__main__":
    main()